pkg/securitypolicy/securitypolicyenforcer_rego.go

```diff
@@ -21,50 +36,491 @@ func init() {
+		policy.compiledModules = compiled
+	} else {
+		return nil, fmt.Errorf("rego compilation failed: %w", err)
+	}
+
+	return policy, nil
+}
+
+func (policy RegoEnforcer) Query(input map[string]interface{}) (rego.ResultSet, error) {
+	store := inmem.NewFromObject(policy.data)
+
+	var buf bytes.Buffer
+	rule := input["name"].(string)
+	query := rego.New(
+		rego.Query(fmt.Sprintf("data.policy.%s", rule)),
+		rego.Compiler(policy.compiledModules),
+		rego.Input(input),
+		rego.Store(store),
+		rego.PrintHook(topdown.NewPrintHook(&buf)))
+
+	ctx := context.Background()
+	results, err := query.Eval(ctx)
+	if err != nil {
+		log.G(ctx).WithError(err).WithFields(logrus.Fields{
+			"Policy": policy.objects,
+		})
+		return results, err
+	}
+
+	output := buf.String()
+	if len(output) > 0 {
+		log.G(ctx).Debug(output)
+	}
+
+	return results, nil
+}
+
+func (policy *RegoEnforcer) EnforceDeviceMountPolicy(target string, deviceHash string) error {
+	policy.mutex.Lock()
+	defer policy.mutex.Unlock()
+
+	input := map[string]interface{}{
+		"name":       "mount_device",
+		"target":     target,
+		"deviceHash": deviceHash,
+	}
+	result, err := policy.Query(input)
+	if err != nil {
+		return err
+	}
+
+	if !result.Allowed() {
+		input_json, err := json.Marshal(input)
+		if err != nil {
+			return fmt.Errorf("Unable to marshal the Rego input data.")
+		}
+
+		return fmt.Errorf("device mount not allowed by policy.\ninput: %s", string(input_json))
+	}
+
+	deviceMap := policy.data["devices"].(map[string]string)
+	if _, found := deviceMap[target]; found {
+		input_json, err := json.Marshal(input)
+		if err != nil {
+			return fmt.Errorf("Unable to marshal the Rego input data.")
+		}
+
+		return fmt.Errorf("device %s already mounted.\ninput: %s", target, string(input_json))
+	}
+
+	deviceMap[target] = deviceHash
+	return nil
+}
+
+func (policy *RegoEnforcer) EnforceOverlayMountPolicy(containerID string, layerPaths []string) error {
+	policy.mutex.Lock()
+	defer policy.mutex.Unlock()
+
+	input := map[string]interface{}{
+		"name":        "mount_overlay",
+		"containerID": containerID,
+		"layerPaths":  layerPaths,
+	}
+	result, err := policy.Query(input)
+	if err != nil {
+		return err
+	}
+
+	if !result.Allowed() {
+		input_json, err := json.Marshal(input)
+		if err != nil {
+			return fmt.Errorf("Unable to marshal the Rego input data.")
+		}
+
+		return fmt.Errorf("overlay mount not allowed by policy.\ninput: %s", string(input_json))
+	}
+
+	// we store the mapping of container ID -> layerPaths for later
+	// use in EnforceCreateContainerPolicy here.
+	containerMap := policy.data["containers"].(map[string]interface{})
+	if _, found := containerMap[containerID]; found {
+		input_json, err := json.Marshal(input)
+		if err != nil {
+			return fmt.Errorf("Unable to marshal the Rego input data.")
+		}
+
+		return fmt.Errorf("container %s already mounted.\ninput: %s", containerID, string(input_json))
```

Comment:
It seems like this "thing didn't work, here's the input to figure it out: %s" workflow is used a bunch, we might be able to just make a custom error type. Naming is up to you:

```go
type RegoError struct {
     Reason, Input string
}

func (re *RegoError) Error() string {
    return fmt.Errorf("%s. Input: %s")
}

func NewRegoError(reason, input string) error {
        return &RegoError{reason, input}
}
```

Reply:
I like this idea. It conflicts with @kevpar's idea of limiting the values in the field and not printing all of input. I'm a fan of the "we don't miss updating an error message" aspect of the current way we print input. But Kevin has a good point about formatting. I'm leaving this til tomorrow to get Matt's thoughts as there are a couple other points on this PR I can't address as I lack knowledge to know "the why" of a couple choices.

Reply:
I'm a huge fan of typed errors. If we want to go down that route, I believe we should have errors for each kind of failure and populate the objects with all the data in input. However, a question worth asking is what we gain for the additional maintenance burden (i.e. having to update all the error types over time as the input changes). Whatever we do, it should be motivated by the needs of the people who will be reading and fixing the errors being reported.

Reply:
I think the more specific we can get about what broke the better in addition to the full input.  Or log the full input once after listing all of the specific errors.

Reply:
I'm in favor of typed errors for specific domains like this, but it sounds like Kevin's suggestion for just printing the devicehash and target in errors was moved to the follow up nit PR. Let's figure out a story for this if we want to go this route there then.